DOCSP-43080: Cursors (#33)
mongodb/docs-c: source/read/cursors.txt

.. _c-cursors:

=========================
Access Data From a Cursor
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: read, results, oplog

Overview
--------

> In this guide, you can learn how to access data from a **cursor** with the

Reviewer: S: avoid "with"
```suggestion
In this guide, you can learn how to access data from a **cursor** by using the
```

{+driver-short+}.

A cursor is a mechanism that returns the results of a read operation in iterable
batches. Because a cursor holds only a subset of documents at any given time,
cursors reduce both memory consumption and the number of requests the driver sends to
the server.

Whenever the {+driver-short+} performs a read operation that returns multiple
documents, it automatically returns those documents in a cursor.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``restaurants`` collection in the ``sample_restaurants``
database from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:atlas:`Get Started with Atlas </getting-started>` guide.

.. _c-cursors-iterate:

Access Cursor Contents Iteratively
----------------------------------

> To iterate over the contents of a cursor, use a while loop. The following example retrieves

Reviewer: S: [nit] I think "while" should be monospaced
```suggestion
To iterate over the contents of a cursor, use a ``while`` loop. The following example retrieves
```

all documents in the ``restaurants`` collection and prints each document by iterating over
the cursor:

.. io-code-block::
   :copyable:

   .. input:: /includes/read/cursors.c
      :start-after: start-cursor-iterate
      :end-before: end-cursor-iterate
      :language: c
      :dedent:

>    .. output::

Reviewer: S: these output documents are pretty long - maybe just show the _id, name, and restaurant_id fields, with ... between _id and name?

      ...

Retrieve Documents Individually
-------------------------------

Retrieve documents from a cursor individually by calling the ``mongoc_cursor_next()`` function.

> When calling the ``mongoc_cursor_next()`` function, the function will iterate the cursor and
> set the ``bson`` parameter to the next document in the cursor.

Reviewer: S: for clarity
```suggestion
This function iterates over the cursor and
sets the ``bson`` parameter to the next document in the cursor.
```

.. literalinclude:: /includes/read/cursors.c
   :start-after: start-cursor-close
   :end-before: end-cursor-close
   :language: c
   :dedent:

Tailable Cursors
----------------

When querying on a :manual:`capped collection </core/capped-collections/>`, you
can use a **tailable cursor** that remains open after the client exhausts the
results in a cursor. To create a tailable cursor on a capped collection,
specify the ``tailable`` and ``awaitData`` options when performing a find operation.

The following example creates a tailable cursor on a capped collection:

.. literalinclude:: /includes/read/cursors.c
   :start-after: start-tailable-cursor
   :end-before: end-tailable-cursor
   :language: c
   :copyable:
   :dedent:

To learn more about tailable cursors and their usage, see the :manual:`Tailable Cursors guide
</core/tailable-cursors/>` in the {+mdb-server+} manual.

Troubleshooting
---------------

"*CursorNotFound* cursor id not valid at server"
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Cursors in MongoDB can timeout on the server if they've been open for
a long time without any operations being performed on them. This can

> lead to a ``CursorNotFound`` exception when you try to iterate through the cursor.

Reviewer: S: could you add another quick sentence about how to troubleshoot this error (I'm guessing just rerun the operation and return a new cursor)?
